src/arith/const_int_bound.cc

```diff
@@ -341,6 +343,18 @@ class ConstIntBoundAnalyzer::Impl
     }
   }
 
+  Entry VisitLeftShift(const CallNode* op) {
+    Entry a = VisitExpr(op->args[0]);
+    Entry b = VisitExpr(op->args[1]);
+
+    // Until C++20, performing a left shift is only well-defined for
+    // positive arguments.  If we have a negative argument, it just
+    // means we couldn't prove that the inputs were positive.
+    a.min_value = std::max(int64_t(0), a.min_value);
+    b.min_value = std::max(int64_t(0), b.min_value);
+    return BinaryOpBoundary(a, b, InfAwareLeftShift);
```

Comment:
if a/b has negative `min_value`, does taking the max narrow the bound? shall we return `Everything` if the bound can't be proved?

Reply:
Hmm, this might depend on the target.  For the `CodegenC` backend, it doesn't narrow the bounds for any legal use of `<<`, because negative arguments are entirely undefined behavior anyways.  Taking the `max` here is a way to express that same constraint.  That is, even if we can't prove that the argument is non-negative, its use in a bitshifting operator provides a constraint.

That said, since my primary goal is to improve simplifications from `ceil_log2`, perhaps it would be better to look for a `ceil(log2(x))` call, and use that directly.  The aggressive optimizations that C++ compilers make based on undefined behavior reasoning are controversial for a reason, and I'd like to avoid introducing similar logic in TVM unless required.

Reply:
Good point.  I had assumed that this was safe, because a similar compile-time check is used for the constant folding in `operator<<`, but in that case the left shift would be entirely removed.  By removing a conditional, but still allowing the left shift to remain, we're asserting that we are handling the left shift identically to how it will be at runtime.

I've updated the handling of left shift to return `Everything` in the case of potentially negative arguments, and added specific handling for `ceil(log2(x))`.